### PR TITLE
New resolver integration

### LINKED ISSUE #92: Implement new coin address resolution
Implement [EIP-2304](https://eips.ethereum.org/EIPS/eip-2304)-compatible coin address resolution.

### LINKED ISSUE #93: Implement contenthash resolution
Implement [EIP-1577](https://eips.ethereum.org/EIPS/eip-1577) compatible content resolution

### PR DESCRIPTION
**Pull request by javiesses**
Closes #92

### CODE REVIEW

File: src/errors/errors.json [hunk 1]
```diff
@@ -286,5 +286,41 @@
...
+   "en": "No contenthash interface",
+   "es": "",
+   "ja": "",
+   "ko": "",
+   "pt": "",
+   "ru": "",
+   "zh": ""
+  }
  }
 }
```
Comment (ilanolkies) [inline src/errors/errors.json]:
```suggestion
}

```

File: src/factories/index.ts [hunk 2]
```diff
@@ -6,6 +6,7 @@ import ReverseRegistrarData from '@rsksmart/rns-reverse/ReverseRegistrarData.jso
 import RSKOwnerData from '@rsksmart/rns-rskregistrar/RSKOwnerData.json';
 import Web3 from 'web3';
 import { AbiItem } from 'web3-utils';
+import ResolverV1Data from '../resolvers/ResolverV1Data.json';
```
Comment (ilanolkies) [inline src/factories/index.ts]: Can we make destructuring assignment here?

File: src/resolutions.ts [hunk 3]
```diff
@@ -239,20 +273,96 @@ export default class extends Composer implements Resolutions {
...
+      this._throw(NO_CONTENTHASH_SET);
+    }
+
+    const decoded = this._contenthashHelper.decodeContenthash(encoded);
+
+    if (!decoded?.protocolType) {
+      this._throw(UNSUPPORTED_CONTENTHASH_PROTOCOL);
+    }
+
+    return decoded!;
```
Comment (ilanolkies) [inline src/resolutions.ts:344]:
```js
    return { decoded!, protocolType };
```

This won't work due to ts. I don't add it as suggestion.

File: src/resolutions.ts [hunk 4]
```diff
@@ -147,14 +163,37 @@ export default class extends Composer implements Resolutions {
...
+      const coinType = this._getCoinTypeFromChainId(chainId);
+      const decodedAddr = await newResolver.methods['addr(bytes32,uint256)'](node, coinType).call();
+
+      if (!decodedAddr || decodedAddr === ZERO_ADDRESS || decodedAddr === '0x') {
+        this._throw(NO_CHAIN_ADDR_RESOLUTION_SET);
+      }
+
+      const buff = Buffer.from(decodedAddr.replace('0x', ''), 'hex');
+
+      return formatsByCoinType[coinType].encoder(buff);
```
Comment (ilanolkies) [inline src/resolutions.ts]: We need to make 0 address check here too.

File: src/resolvers/ProxyAdminData.json [hunk 5]
```diff
@@ -0,0 +1 @@
...50505050565b61090b6106e5565b61091457600080fd5b61091d816109e8565b50565b60008060608373ffffffffffffffffffffffffffffffffffffffff1660405180807ff851a440000000000000000000000000000000000000000000000000000000008152506004019050600060405180830381855afa9150503d80600081146109a5576040519150601f19603f3d011682016040523d82523d6000602084013e6109aa565b606091505b5091509150816109b957600080fd5b8080602001905160208110156109ce57600080fd5b810190808051906020019092919050505092505050919050565b600073ffffffffffffffffffffffffffffffffffffffff168173ffffffffffffffffffffffffffffffffffffffff161415610a2257600080fd5b8073ffffffffffffffffffffffffffffffffffffffff166000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff167f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e060405160405180910390a3806000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055505056fea265627a7a72315820f9fa5f4056a82fc6924be0114c6bf0ad4985b381d5d7a37a096d8f1ff1b2653264736f6c63430005100032","address":{"rskMainnet":"0x30569C0999f0C7fD7be9957ac4B4dEe803CAc293","rskTestnet":"0xbb71C17b28BAF2B9e7f5b31660eCF758113b3fEf"}}
```
Comment (ilanolkies) [inline src/resolvers/ProxyAdminData.json]:
```suggestion
{"abi":[{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"previousOwner","type":"address"},{"indexed":true,"internalType":"address","name":"newOwner","type":"address"}],"name":"OwnershipTransferred","type":"event"},{"constant":true,"inputs":[],"name":"isOwner","outputs":[{"internalType":"bool","name":"","type":"bool"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"owner","outputs":[{"internalType":"address","name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[],"name":"renounceOwnership","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":false,"inputs":[{"internalType":"address","name":"newOwner","type":"address"}],"name":"transferOwnership","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"internalType":"contract AdminUpgradeabilityProxy","name":"proxy","type":"address"}],"name":"getProxyImplementation","outputs":[{"internalType":"address","name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"internalType":"contract AdminUpgradeabilityProxy","name":"proxy","type":"address"}],"name":"getProxyAdmin","outputs":[{"internalType":"address","name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"internalType":"contract AdminUpgradeabilityProxy","name":"proxy","type":"address"},{"internalType":"address","name":"newAdmin","type":"address"}],"name":"changeProxyAdmin","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":false,"inputs":[{"internalType":"contract AdminUpgradeabilityProxy","name":"proxy","type":"address"},{"internalType":"address","name":"implementation","type":"address"}],"name":"upgrade","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":false,"inputs":[{"internalType":"contract AdminUpgradeabilityProxy","name":"proxy","type":"address"},{"internalType":"address","name":"implementation","type":"address"},{"internalType":"bytes","name":"data","type":"bytes"}],"name":"upgradeAndCall","outputs":[],"payable":true,"stateMutability":"payable","type":"function"}],"bytecode":"0x6080604052336000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055506000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16600073ffffffffffffffffffffffffffffffffffffffff167f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e060405160405180910390a3610b15806100cf6000396000f3fe6080604052600436106100865760003560e01c80638f32d59b116100595780638f32d59b146101fb5780639623609d1461022a57806399a88ec414610325578063f2fde38b14610396578063f3b7dead146103e757610086565b8063204e1c7a1461008b578063715018a61461011c5780637eff275e146101335780638da5cb5b146101a4575b600080fd5b34801561009757600080fd5b506100da600480360360208110156100ae57600080fd5b81019080803573ffffffffffffffffffffffffffffffffffffffff169060200190929190505050610478565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b34801561012857600080fd5b50610131610540565b005b34801561013f57600080fd5b506101a26004803603604081101561015657600080fd5b81019080803573ffffffffffffffffffffffffffffffffffffffff169060200190929190803573ffffffffffffffffffffffffffffffffffffffff169060200190929190505050610610565b005b3480156101b057600080fd5b506101b96106bc565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b34801561020757600080fd5b506102106106e5565b604051808215151515815260200191505060405180910390f35b6103236004803603606081101561024057600080fd5b81019080803573ffffffffffffffffffffffffffffffffffffffff169060200190929190803573ffffffffffffffffffffffffffffffffffffffff1690602001909291908035906020019064010000000081111561029d57600080fd5b8201836020820111156102af57600080fd5b803590602001918460018302840111640100000000831117156102d157600080fd5b91908080601f016020809104026020016040519081016040528093929190818152602001838380828437600081840152601f19601f82011690508083019250505050505050919291929050505061073c565b005b34801561033157600080fd5b506103946004803603604081101561034857600080fd5b81019080803573ffffffffffffffffffffffffffffffffffffffff169060200190929190803573ffffffffffffffffffffffffffffffffffffffff169060200190929190505050610857565b005b3480156103a257600080fd5b506103e5600480360360208110156103b957600080fd5b81019080803573ffffffffffffffffffffffffffffffffffffffff169060200190929190505050610903565b005b3480156103f357600080fd5b506104366004803603602081101561040a57600080fd5b81019080803573ffffffffffffffffffffffffffffffffffffffff169060200190929190505050610920565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b60008060608373ffffffffffffffffffffffffffffffffffffffff1660405180807f5c60da1b000000000000000000000000000000000000000000000000000000008152506004019050600060405180830381855afa9150503d80600081146104fd576040519150601f19603f3d011682016040523d82523d6000602084013e610502565b606091505b50915091508161051157600080fd5b80806020019051602081101561052657600080fd5b810190808051906020019092919050505092505050919050565b6105486106e5565b61055157600080fd5b600073ffffffffffffffffffffffffffffffffffffffff166000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff167f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e060405160405180910390a360008060006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff160217905550565b6106186106e5565b61062157600080fd5b8173ffffffffffffffffffffffffffffffffffffffff16638f283970826040518263ffffffff1660e01b8152600401808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff168152602001915050600060405180830381600087803b1580156106a057600080fd5b505af11580156106b4573d6000803e3d6000fd5b505050505050565b60008060009054906101000a900473ffffffffffffffffffffffffffffffffffffffff16905090565b60008060009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff1614905090565b6107446106e5565b61074d57600080fd5b8273ffffffffffffffffffffffffffffffffffffffff16634f1ef2863484846040518463ffffffff1660e01b8152600401808373ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200180602001828103825283818151815260200191508051906020019080838360005b838110156107ed5780820151818401526020810190506107d2565b50505050905090810190601f16801561081a5780820380516001836020036101000a031916815260200191505b5093505050506000604051808303818588803b15801561083957600080fd5b505af115801561084d573d6000803e3d6000fd5b5050505050505050565b61085f6106e5565b61086857600080fd5b8173ffffffffffffffffffffffffffffffffffffffff16633659cfe6826040518263ffffffff1660e01b8152600401808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff168152602001915050600060405180830381600087803b1580156108e757600080fd5b505af11580156108fb573d6000803e3d6000fd5b505050505050565b61090b6106e5565b61091457600080fd5b61091d816109e8565b50565b60008060608373ffffffffffffffffffffffffffffffffffffffff1660405180807ff851a440000000000000000000000000000000000000000000000000000000008152506004019050600060405180830381855afa9150503d80600081146109a5576040519150601f19603f3d011682016040523d82523d6000602084013e6109aa565b606091505b5091509150816109b957600080fd5b8080602001905160208110156109ce57600080fd5b810190808051906020019092919050505092505050919050565b600073ffffffffffffffffffffffffffffffffffffffff168173ffffffffffffffffffffffffffffffffffffffff161415610a2257600080fd5b8073ffffffffffffffffffffffffffffffffffffffff166000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff167f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e060405160405180910390a3806000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055505056fea265627a7a72315820f9fa5f4056a82fc6924be0114c6bf0ad4985b381d5d7a37a096d8f1ff1b2653264736f6c63430005100032","address":{"rskMainnet":"0x30569C0999f0C7fD7be9957ac4B4dEe803CAc293","rskTestnet":"0xbb71C17b28BAF2B9e7f5b31660eCF758113b3fEf"}}

```

File: test/local/setAddr.test.ts [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+  });
+
+  describe('definitive resolver', () => {
+    beforeEach(async () => {
+      ({ registry } = await deployDefinitiveResolver());
+
+      rns = getRNSInstance(blockchainApiInstance, registry);
+    });
+
+    it('should set an address with definitive resolver', async () => {
```
Comment (ilanolkies) [inline test/local/setAddr.test.ts:167]: Can we add a test to set the address back to 0?

File: test/local/setChainAddr.test.ts [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        const txReceipt = await rns.setAddr('alice.rsk', ethAddr, ChainId.ETHEREUM, { from });
+
+        const tx = await web3.eth.getTransaction(txReceipt.transactionHash);
+
+        expect(tx.from).toEqual(from);
+      });
+    });
+  });
+
+  describe('definitive resolver', () => {
```
Comment (ilanolkies) [inline test/local/setChainAddr.test.ts:163]: Same, can we add a test to set the address back to 0?

File: test/local/setContenthash.test.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  it('should set onion contenthash', async () => {
+    const onionHash = 'onion://zqktlwi4fecvo6ri';
+    await shouldEncodeProperly(onionHash);
+  });
+
+  it('should set onion3 contenthash', async () => {
+    const onionHash = 'onion3://p53lf57qovyuvwsc6xnrppyply3vtqm7l6pcobkmyqsiofyeznfu5uqd';
+    await shouldEncodeProperly(onionHash);
+  });
+
```
Comment (ilanolkies) [inline test/local/setContenthash.test.ts:64]: Here too, can we add a test to set the content back to null?

**Review by ilanolkies — CHANGES_REQUESTED** [review]
LRGTM. Minor changes.

**Review by ilanolkies — APPROVED** [review]
